packages/webapp/src/components/Forms/Sale/index.jsx

```diff
@@ -1,85 +1,244 @@
+  // Reformat selector to match component format
+  // TODO: match component to selector format
+  const existingSales = sale?.crop_variety_sale.reduce(
+    (acc, cur) => ({
+      ...acc,
+      [cur.crop_variety_id]: {
+        crop_variety_id: cur.crop_variety_id,
+        quantity: cur.quantity,
+        quantity_unit: cur.quantity_unit,
+        sale_value: cur.sale_value,
+      },
+    }),
+    {},
+  );
+  // FilterPillSelect details
+  const STATUS = 'STATUS';
+  const filterRef = useRef({});
+  const filter = {
+    filterKey: STATUS,
+    options: cropVarietyOptions.map((cvs) => ({
+      value: cvs.label,
+      default: existingSales?.[cvs.value] ? true : false,
+      label: cvs.label,
+    })),
+  };
+  // Unique hook form names
   const { t } = useTranslation();
+  const SALE_DATE = 'sale_date';
+  const SALE_CUSTOMER = 'customer_name';
+  const CHOSEN_VARIETIES = 'crop_variety_sale';
+  let CROP_VARIETY_VALUES = {};
+  cropVarietyOptions.forEach((cv) => {
+    CROP_VARIETY_VALUES[cv.value] = {
+      CROP_VARIETY_ID: `${CHOSEN_VARIETIES}.${cv.value}.crop_variety_id`,
+      SALE_VALUE: `${CHOSEN_VARIETIES}.${cv.value}.sale_value`,
+    };
+  });
+
+  const {
+    register,
+    handleSubmit,
+    watch,
+    getValues,
+    setValue,
+    control,
+    formState: { errors, isValid },
+  } = useForm({
+    mode: 'onChange',
+    shouldUnregister: true,
+    defaultValues: {
+      [SALE_DATE]: getDateInputFormat(sale?.sale_date) || getDateInputFormat(),
+      customer_name: sale?.customer_name,
+      crop_variety_sale: existingSales ?? undefined,
+    },
+  });
+
+  const saleDateRegister = register(SALE_DATE, { required: true });
+  const saleCustomerRegister = register(SALE_CUSTOMER, { required: true });
+  // FilterPillSelect does not support register yet, so const definition not needed
+  register(CHOSEN_VARIETIES, { required: true });
+
+  const [isDirty, setIsDirty] = useState(false);
+  const [filterState, setFilterState] = useState(filter);
+  const [isFilterValid, setIsFilterValid] = useState(true);
+  const [chosenOptions, setChosenOptions] = useState([]);
+  const [cropVarietyRegisters, setCropVarietyRegisters] = useState({});
+
+  // Run once and do not assess filter validity -- needed for edit sale
+  useEffect(() => {
+    const activeOptions = cropVarietyOptions.filter((cvs) =>
+      existingSales?.[cvs.value] ? true : false,
+    );
+    activeOptions.length ?? setIsFilterValid(true);
+    // Input does not support registering like Unit, dynamically register here
+    let dynamicRegister = {};
+    activeOptions.forEach((option) => {
+      dynamicRegister[option.value] = {
+        cropVarietyIdRegister: register(CROP_VARIETY_VALUES[option.value].CROP_VARIETY_ID, {
+          required: true,
+          value: option.value,
+        }),
+        saleValueRegister: register(CROP_VARIETY_VALUES[option.value].SALE_VALUE, {
+          required: true,
+          valueAsNumber: true,
+        }),
+      };
+    });
```

Comment:
I think these can be done in one loop, and you also can create a function to return activeOptions and dynamicRegister. This looks almost the same as what's done in the useEffect below.

Reply:
I agree! haha It does bother me that I couldn't find a nice solution to that yet.

/EditSale and /AddSale are almost exactly the same too I would love to combine. This component still needs work to be clean code.